lib/lightning/pipeline/runner.ex

```diff
@@ -170,46 +172,92 @@ defmodule Lightning.Pipeline.Runner do
+
+  The following scenarios are handled:
+  1. Reads the file content of the result's final state.
+  2. Attempts to decode the file content as JSON.
+  3. Processes the decoded content, linking it to the provided run.
+
+  If the file doesn't exist or there's a JSON decoding error:
+  - The error is logged.
+  - A nil value is set for the output dataclip linked to the run.
+  - An error tuple is returned.
+
+  Success scenario:
+  - The dataclip is updated with the decoded content and linked to the run.
+
+  Returns:
+  - The updated run with the linked dataclip in the case of success.
+  - An error tuple in the case of failure.
+
   """
   @spec create_dataclip_from_result(
           result :: Lightning.Runtime.Result.t(),
           run :: Invocation.Run.t()
         ) ::
           {:ok, Invocation.Dataclip.t()} | {:error, any}
-  def create_dataclip_from_result(
-        %Lightning.Runtime.Result{} = result,
-        run
-      ) do
-    with {:ok, data} <- File.read(result.final_state_path),
-         {:ok, body} <- Jason.decode(data) do
-      job = Lightning.Repo.preload(run.job, :workflow)
-
-      Invocation.update_run(run, %{
-        output_dataclip: %{
-          project_id: job.workflow.project_id,
-          type: :run_result,
-          body: scrub_result(body)
-        }
-      })
-    else
-      res = {:error, %Jason.DecodeError{position: pos}} ->
-        Logger.info(
-          "Got JSON decoding error when trying to parse: #{result.final_state_path}:#{pos}"
-        )
-
-        res
+  def create_dataclip_from_result(%Lightning.Runtime.Result{} = result, run) do
+    result.final_state_path
+    |> read_file_content()
+    |> decode_file_content()
+    |> process_decoded_content(run)
+  end
 
-      res = {:error, err} ->
-        Logger.info(
-          "Got unexpected result while saving the resulting state from a Run:\n#{inspect(err)}"
-        )
+  defp read_file_content(file_path) do
+    case File.read(file_path) do
+      {:ok, data} -> {:ok, data}
+      error -> error
+    end
+  end
 
-        res
+  defp decode_file_content({:ok, data}) do
+    case Jason.decode(data) do
+      {:ok, body} -> {:ok, body}
+      error -> error
     end
   end
 
+  defp decode_file_content(error), do: error
+
+  defp process_decoded_content({:ok, body}, run) do
+    job = Lightning.Repo.preload(run.job, :workflow)
+
+    Invocation.update_run(run, %{
+      output_dataclip: %{
+        project_id: job.workflow.project_id,
+        type: :run_result,
+        body: scrub_result(body)
+      }
+    })
+  end
+
+  defp process_decoded_content(
+         {:error, %Jason.DecodeError{position: pos} = error},
+         run
+       ) do
+    Logger.info(
+      "Got JSON decoding error when trying to parse: #{run.final_state_path}:#{pos}"
+    )
+
+    run
+    |> Repo.preload(:output_dataclip)
+    |> Invocation.update_run(%{
+      output_dataclip: nil
+    })
+
+    {:error, error}
```

Comment:
Is this the actual fix here? If so then rather put these calls inside the `else` block for the `with` in the original implementation no?